### PR TITLE
fix: impose a memory limit

### PR DESCRIPTION
**Pull request by onbjerg**
Some Foundry tests might access memory at very high indexes that would not otherwise be possible due to gas constraints. A solution would be to lower the amount of gas tests can use, but ideally we'd want this to be as uncapped as possible to allow for deployment of big test contracts.

This imposes a memory limit of 16 megabytes for the interpreter which (I think?) should be more than enough. Let me know what you think/if there is a better solution. Could also be feature-gated

### CODE REVIEW

**Comment by rakita** [comment]
I feal like this is not the best way to do it, i still need to check how geth and oe bahave in this case just to be on same page with them.

File: crates/revm/src/interpreter/memory.rs [hunk 1]
```diff
@@ -5,6 +5,12 @@ use core::{
 };
 use primitive_types::U256;
 
+/// The absolute memory limit (16 megabytes).
+///
+/// Prevents panicks if the gas limit is really high and memory access/stores that are otherwise
+/// not realistic in a live environment become possible.
+pub static MEMORY_LIMIT: usize = 2usize.pow(24);
```
[the remark this answers is not in the document]
Comment (onbjerg) [reply crates/revm/src/interpreter/memory.rs:12]: I think 2^32 is too much for Foundry (that would mean that fuzz tests could RNG allocate 4 gigs of memory!), so this should likely be configurable instead

File: crates/revm/src/instructions/macros.rs [hunk 2]
```diff
@@ -53,6 +53,10 @@ macro_rules! memory_resize {
         if let Some(new_size) =
             crate::interpreter::memory::next_multiple_of_32(offset.saturating_add(len))
         {
+            if new_size > crate::interpreter::memory::MEMORY_LIMIT {
+                return Return::FatalNotSupported;
```
Comment (hrkrshnn) [inline crates/revm/src/instructions/macros.rs:57]: The behaviour recommended in EIP-1985 is to go out of gas.

Note: solidity has runtime code for reverting if your memory allocation extends the free memory pointer more than `2**64 -1`.

File: crates/revm/src/interpreter/memory.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (rakita) [reply crates/revm/src/interpreter/memory.rs:12]: Thank for a share, this seems like good default limits, but for foundry fuzzing it is fine to introduce feature gate and set a lot smaller value.

**Comment by alcuadrado** [comment]
Wouldn't this fit better in `OverrideSpec`? Idk if Foundry uses an `Inspector` though, which is needed to use `OverrideSpec`, but I guess it does.

**Comment by onbjerg** [comment]
> Wouldn't this fit better in `OverrideSpec`? Idk if Foundry uses an `Inspector` though, which is needed to use `OverrideSpec`, but I guess it does.

Dragan previously talked about removing the override spec thing so I opted not to put it in there

**Comment by onbjerg** [comment]
@rakita This is now hidden behind the `memory_limit` feature and is configurable through `CfgEnv`. Let me know your thoughts

File: crates/revm/src/interpreter/memory.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (rakita) [reply crates/revm/src/interpreter/memory.rs:12]: nit, not used.

[the remark this answers is not in the document]
Comment (onbjerg) [reply crates/revm/src/interpreter/memory.rs:12]: oh, right, mb

**Pull request merged**